### PR TITLE
Allow user to delete task by name

### PR DESCRIPTION
**Pull request by Skaty**
Author: @AHaliq

### CODE REVIEW

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 1]
```diff
@@ -10,11 +10,19 @@
  */
 public class StringUtil {
     public static boolean containsIgnoreCase(String source, String query) {
-        String[] split = source.toLowerCase().split("\\s+");
-        List<String> strings = Arrays.asList(split);
+        List<String> strings = listifyString(source);
```
Comment (advaypal) [inline src/main/java/seedu/address/commons/util/StringUtil.java:13]: I'm not sure about this, but does this break SLAP?

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 2]
```diff
@@ -13,35 +13,54 @@
...
             + "Example: " + COMMAND_WORD + " 1";
 
     public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
 
     public final int targetIndex;
+    public final String targetName;
 
     public DeleteCommand(int targetIndex) {
         this.targetIndex = targetIndex;
+        this.targetName = null;
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:27]: Instead of setting it to null here, set the default values of both at the top(outside constructor), and then initialise only the required one here

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 3]
```diff
@@ -13,35 +13,54 @@
...
+    }
+
+    public DeleteCommand(String name) {
+        this.targetIndex = -1;
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:31]: The -1 value here is not being used anywhere. I think it's better not to initialise it at all

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 4]
```diff
@@ -13,35 +13,54 @@
...
+            }
 
-        ReadOnlyTask taskToDelete = lastShownList.get(targetIndex - 1);
+            taskToDelete = lastShownList.get(targetIndex - 1);
 
-        try {
-            model.deleteTask(taskToDelete);
-        } catch (TaskNotFoundException pnfe) {
-            assert false : "The target task cannot be missing";
+            try {
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:59]: You can have a variable to represent the task name, and then put this try catch block outside the if-else condition(since it's currently in both). The variable can be updated within the if else condition. This prevents unnecessary duplication :)

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 5]
```diff
@@ -13,35 +13,54 @@
...
     public CommandResult execute() {
+        ReadOnlyTask taskToDelete;
 
-        UnmodifiableObservableList<ReadOnlyTask> lastShownList = model.getFilteredTaskList();
+        if(this.targetName != null) {
+            try {
+                taskToDelete = model.deleteTask(this.targetName);
+            } catch (TaskNotFoundException pnfe) {
+                indicateAttemptToExecuteIncorrectCommand();
+                return new CommandResult(Messages.MESSAGE_INVALID_TASK_NAME_SEARCH);
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:45]: I believe this should be statment? (looking at the rest of the code):
assert false : "The target task cannot be missing"; and not MESSAGE_INVALID_TASK_NAME_SEARCH

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 6]
```diff
@@ -128,12 +129,24 @@ private Command prepareAdd(String args){
...
         Optional<Integer> index = parseIndex(args);
+        // check if argument is index
+
         if(!index.isPresent()){
-            return new IncorrectCommand(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE));
+            String nameMatch = args.length() > 0 ? args.substring(1) : args;
+            // remove white space from command delimiter
+
+            if(nameMatch.contentEquals("")) {
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/parser/Parser.java:138]: Not too much of an issue, but why contentEquals over equals?

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 7]
```diff
@@ -128,12 +129,24 @@ private Command prepareAdd(String args){
     private Command prepareDelete(String args) {
 
         Optional<Integer> index = parseIndex(args);
+        // check if argument is index
+
         if(!index.isPresent()){
-            return new IncorrectCommand(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE));
+            String nameMatch = args.length() > 0 ? args.substring(1) : args;
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/parser/Parser.java:135]: Could you explain this to me?

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 8]
```diff
@@ -128,12 +129,24 @@ private Command prepareAdd(String args){
...
+                        String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE));
+            }
+            // invalid delete command if no arguments
+
+            return new DeleteCommand(nameMatch);
+            // create delete by name
         }
 
         return new DeleteCommand(index.get());
+        // create delete by index
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/parser/Parser.java:149]: This line gets executed even if there is no index and no name. You need to wrap this in a check

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 9]
```diff
@@ -177,16 +190,26 @@ private Command prepareSelect(String args) {
      * @return the prepared command
      */
     private Command prepareFind(String args) {
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/parser/Parser.java:192]: Why have you changed this method and added a new one? What's the point?

File: src/main/java/seedu/address/model/AddressBook.java [hunk 10]
```diff
@@ -51,6 +51,26 @@ public static ReadOnlyAddressBook getEmptyAddressBook() {
         return tasks.getInternalList();
     }
 
+    /**
```
Comment (advaypal) [inline src/main/java/seedu/address/model/AddressBook.java:54]: Where is this method being used in the rest of the code?

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 11]
```diff
@@ -13,35 +13,54 @@
...
 
 
     @Override
     public CommandResult execute() {
+        ReadOnlyTask taskToDelete;
 
-        UnmodifiableObservableList<ReadOnlyTask> lastShownList = model.getFilteredTaskList();
+        if(this.targetName != null) {
+            try {
+                taskToDelete = model.deleteTask(this.targetName);
```
Comment (advaypal) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:42]: taskToDelete = model.deleteTask(this.targetName) ?

This doesn't make a lot of sense to me. You're deleting a task and setting that task as the task to delete?

File: src/main/java/seedu/address/model/Model.java [hunk 12]
```diff
@@ -20,6 +20,7 @@
 
     /** Deletes the given task. */
     void deleteTask(ReadOnlyTask target) throws UniqueTaskList.TaskNotFoundException;
+    ReadOnlyTask deleteTask(String target) throws UniqueTaskList.TaskNotFoundException;
```
Comment (advaypal) [inline src/main/java/seedu/address/model/Model.java:23]: What's the point of this? We already have a deleteTask.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 13]
```diff
@@ -71,6 +71,13 @@ public synchronized void deleteTask(ReadOnlyTask target) throws TaskNotFoundExce
     }
 
     @Override
+    public synchronized ReadOnlyTask deleteTask(String target) throws TaskNotFoundException {
```
Comment (advaypal) [inline src/main/java/seedu/address/model/ModelManager.java:74]: Same question, Why is this needed?

File: src/main/java/seedu/address/model/ModelManager.java [hunk 14]
```diff
@@ -91,7 +98,7 @@ public void updateFilteredListToShowAll() {
 
     @Override
     public void updateFilteredTaskList(Set<String> keywords){
```
Comment (advaypal) [inline src/main/java/seedu/address/model/ModelManager.java:100]: What is this NameSubstringQualifier for? How does it tie in with delete?

**Review by advaypal — CHANGES_REQUESTED** [review]
I've made a bunch of comments. You'll also need to add more tests before we can merge this in.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 2 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:27]: Both are final variables (following the design of the original source code) thus upon first initialization you cant change its value any more. And depending on the constructor (we have two now) one has to be set to a valid value. Thats why its not initialized upon definition. If you were to read original source they did the same thing except with only the targetIndex variable. It is self explanatory from seeing the diff.

File: src/main/java/seedu/address/commons/util/StringUtil.java (hunk 1 above)
Comment (AHaliq) [reply src/main/java/seedu/address/commons/util/StringUtil.java:13]: Simply abstracting out a reusable method. I dont see how that breaks slap, unless ud want to wrap the remaining lines of code in the body into another method unecessarily.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 3 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:31]: See prev comment

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 11 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:42]: If u were to read the whole code block, you can see taskToDelete is used for success message. Even in the original source as well.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 5 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:45]: In this scenario the task CAN be missing as the user can specify names that are invalid and searching for tasks is done from command not in parser. Unlike index which was processed from parser.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 4 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:59]: They are two distinct errors and conditions for errors, we could lump them into one but im simply following the spirit of the source code to differentiate all different failure types with their respective messages.

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 7 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:135]: That the execute command checks if its invoked from its index as argument constructor vs name as argument constructor?

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 6 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:138]: No difference really. Content equals works for other character streams as well.

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 8 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:149]: I think its perfectly fine as going through that block with no index and name will still produce the desired output of informing the user of invalid format. Which is what we want isnt it?

File: src/main/java/seedu/address/model/ModelManager.java (hunk 14 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/ModelManager.java:100]: This is from find command. "Delete branch" should be merged with "find branch" before merged to master. See who created the prs. Im assuming this is what you two agreed upon.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 13 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/ModelManager.java:74]: If you can read, deleting the task with name argument. Method overloading. Alternative interfaces to do the same execution rather than executing outside the class of responsibility.

File: src/main/java/seedu/address/model/Model.java (hunk 12 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/Model.java:23]: See comment below

File: src/main/java/seedu/address/model/AddressBook.java (hunk 10 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/AddressBook.java:54]: Use eclipse 'go to'

Comment (AHaliq) [inline src/main/java/seedu/address/model/AddressBook.java]: Eclipse*

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 9 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:192]: As ive abstracted out the construction of the keyword list. It could be useful for future commands using keywords.

**Comment by AHaliq** [comment]
@advaypal ready for review

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 3 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:31]: You can delete the this.targetIndex = -1. It's not necessary.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 2 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:27]: Ok no need to do that, but see the next comment.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 11 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:42]: taskToDelete is the task to be deleted, it's not created just because of the error message. You are setting the variable which represents the task to be deleted in the future, to a task that has already been deleted. That's an example of terrible variable naming. Furthermore, you created a new deleteTask method just to return the deleted task, but you don't use that information anywhere. Please replace this line of code with simply model.deleteTask(this.name) and remove that extra deleteTask method you created in the model.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 5 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:45]: In this scenario the task CANNOT be missing, because the user cannot delete a task that does not exist.

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 4 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:59]: No, they are not distinct. See the comment above.

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 7 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/parser/Parser.java:135]: Where have you changed the call of prepareDelete to take in the name in args?

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 6 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/parser/Parser.java:138]: ok

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 8 above)
Comment (advaypal) [reply src/main/java/seedu/address/logic/parser/Parser.java:149]: Wouldn't an error be caused if this line is executed? Because taskToDelete = lastShownList.get(targetIndex - 1); is not wrapped in a try catch

File: src/main/java/seedu/address/model/AddressBook.java (hunk 10 above)
Comment (advaypal) [inline src/main/java/seedu/address/model/AddressBook.java]: I don't think I can do that on GitHub.

Comment (advaypal) [reply src/main/java/seedu/address/model/AddressBook.java:54]: This method isn't needed, you can find task by name using the methods already in the codebase. Furthermore, what if finding by a particular name returns multiple tasks? Your method doesn't take care of that.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 13 above)
Comment (advaypal) [reply src/main/java/seedu/address/model/ModelManager.java:74]: This method is not needed. Have commented on this in the usage above.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 15]
```diff
@@ -71,6 +71,13 @@ public synchronized void deleteTask(ReadOnlyTask target) throws TaskNotFoundExce
     }
 
     @Override
+    public synchronized ReadOnlyTask deleteTask(String target) throws TaskNotFoundException {
+        ReadOnlyTask capture = addressBook.getTask(target);
```
Comment (advaypal) [inline src/main/java/seedu/address/model/ModelManager.java:75]: See how the original source was getting tasks. Creating your own method here is a waste of time.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 14 above)
Comment (advaypal) [reply src/main/java/seedu/address/model/ModelManager.java:100]: No, we did not agree on this. @Skaty was under the assumption that you wouldn't mix the two together. Please do not do that in the future, as discussed earlier.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 16]
```diff
@@ -150,4 +157,25 @@ public String toString() {
...
 
+    private class NameSubstringQualifier implements Qualifier {
+        private Set<String> nameKeyWords;
+
+        NameSubstringQualifier(Set<String> nameKeyWords) {
+            this.nameKeyWords = nameKeyWords;
+        }
+
+        @Override
+        public boolean run(ReadOnlyTask task) {
```
Comment (advaypal) [inline src/main/java/seedu/address/model/ModelManager.java]: There are two 'runs' and two 'toStrings'. Please delete the duplicated methods.

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 17]
```diff
@@ -227,10 +227,11 @@ public void execute_list_showsAllTasks() throws Exception {
      */
     private void assertIncorrectIndexFormatBehaviorForCommand(String commandWord, String expectedMessage) throws Exception {
         assertCommandBehavior(commandWord , expectedMessage); //index missing
-        assertCommandBehavior(commandWord + " +1", expectedMessage); //index should be unsigned
-        assertCommandBehavior(commandWord + " -1", expectedMessage); //index should be unsigned
-        assertCommandBehavior(commandWord + " 0", expectedMessage); //index cannot be 0
-        assertCommandBehavior(commandWord + " not_a_number", expectedMessage);
+        //assertCommandBehavior(commandWord + " +1", expectedMessage); //index should be unsigned
```
Comment (advaypal) [inline src/test/java/seedu/address/logic/LogicManagerTest.java:230]: Why have you commented these tests out?

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 3 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:31]: Error:(32, 5) java: variable targetIndex might not have been initialized

when removed

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 4 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:59]: cases for delete command
1) argument is a valid index
2) argument is an invalid index (index > length)
3) argument is a valid string
4) argument is an invalid string (no task has the name)
2 and 4 are distinct

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 5 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java:45]: to determine if a task is missing you would have to process the string, to do that you need the model object, thus it is inappropriate to pre process the argument in parser. Thus the processing is done in the DeleteCommand class, thus it can be missing. Please advise next course of action rather than simply imposing a design.

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 7 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:135]: i didnt? prepareDelete(arg) takes an argument, this can be both index and task name. processing is done in the code body. Im not sure why you're confused here.

File: src/main/java/seedu/address/logic/parser/Parser.java (hunk 8 above)
Comment (AHaliq) [reply src/main/java/seedu/address/logic/parser/Parser.java:149]: im not getting errors, follow the stack of throw catches to convince yourself

File: src/main/java/seedu/address/model/AddressBook.java (hunk 10 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/AddressBook.java:54]: you can find task by index only. theres no implementation for getting the readable task object via name. unless ive overseen one. in that case please explain, dont just state without explaining.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 13 above)
Comment (AHaliq) [reply src/main/java/seedu/address/model/ModelManager.java:74]: Its delete by name, not by index, please provide alternative i cant find implementations of by name
